test/hotspot/gtest/utilities/test_resourceHash.cpp

```diff
@@ -287,3 +288,163 @@ TEST_VM_F(GenericResourceHashtableTest, bad_hash_no_rm) {
+  s->increment_refcount();
+  _simple_test_table.put(s, 55);
+  ASSERT_EQ(s->refcount(), s_orig_count + 1) << "refcount should be incremented in table";
+
+  // Deleting this value from a hashtable
+  _simple_test_table.remove(s);
+  // Now decrement the refcount for s since it's no longer in the table.
+  s->decrement_refcount();
+  ASSERT_EQ(s->refcount(), s_orig_count) << "refcount should be same as start";
+}
+
+TEST_VM_F(ResourceHashtableDeleteTest, simple_delete) {
+  TempNewSymbol s = SymbolTable::new_symbol("abcdefg_simple");
+  int s_orig_count = s->refcount();
+  // Need to increment a Symbol* when you keep it in a table.
+  s->increment_refcount();
+  _simple_test_table.put(s, 55);
+  ASSERT_EQ(s->refcount(), s_orig_count + 1) << "refcount should be incremented in table";
+
+  // Use unlink to remove the matching (or all) values from the table.
+  SimpleDeleter deleter;
+  _simple_test_table.unlink(&deleter);
+  ASSERT_EQ(s->refcount(), s_orig_count) << "refcount should be same as start";
+}
+
+TEST_VM_F(ResourceHashtableDeleteTest, value_remove) {
+  TempNewSymbol s = SymbolTable::new_symbol("abcdefg");
+  int s_orig_count = s->refcount();
+  {
+    TestValue tv(s);
+    // If you use a Symbol* as key in the ResourceHashtable, you must increment the
+    // refcount outside the hashtable functions.
+    s->increment_refcount();
+    _test_table.put(s, tv);
+    ASSERT_EQ(s->refcount(), s_orig_count + 3) << "refcount incremented";
+  }
+  ASSERT_EQ(s->refcount(), s_orig_count + 2) << "refcount not copied";
+
+  // Deleting this value from a hashtable calls the destructor!
+  _test_table.remove(s);
+  // Now decrement the refcount for s since it's no longer in the table.
+  s->decrement_refcount();
+  // Removal should make the refcount be the original refcount.
+  ASSERT_EQ(s->refcount(), s_orig_count) << "refcount now decremented";
+}
+
+TEST_VM_F(ResourceHashtableDeleteTest, value_delete) {
+  TempNewSymbol d = SymbolTable::new_symbol("defghijklmnop");
+  int d_orig_count = d->refcount();
+  {
+    TestValue tv(d);
+    // On the other hand, if the Key Symbol* is already a member of the entry, it's not strictly
+    // necessary to increment the refcount on the symbol.
+    _test_table.put(d, tv);
+    ASSERT_EQ(d->refcount(), d_orig_count + 2) << "refcount incremented";
+  }
+  ASSERT_EQ(d->refcount(), d_orig_count + 1) << "refcount incremented";
+  Deleter deleter;
+  _test_table.unlink(&deleter);
+  ASSERT_EQ(d->refcount(), d_orig_count) << "refcount should be as we started";
+}
+
+TEST_VM_F(ResourceHashtableDeleteTest, check_delete_ptr) {
+  TempNewSymbol s = SymbolTable::new_symbol("abcdefg_ptr");
+  int s_orig_count = s->refcount();
+  {
+    TestValue* tv = new TestValue(s);
+    // If you use a Symbol* as key in the ResourceHashtable, you must increment the
+    // refcount outside the hashtable functions.
+    s->increment_refcount();
+    _ptr_test_table.put(s, tv);
```

Comment:
I am not sure why it's necessary to do s->increment_refcount() here and key->decrement_refcount() in PtrDeleter. We know that the symbol will be kept alive by the TestValue. These two inc/dec calls are done where we know that the TestValue is alive (after instantiation, and before deallocation). By removing these calls, we can reduce the chance of mishandling the refcount.

Reply:
Yes, I was trying to show that you may have to manipulate the refcount in the Key if it's not kept alive by being the same in the Value. In this case it is.  Adding the SimpleDeleter to show that does a better job of that though.